service-deployment/bootstrap/hadoop-service/hadoop-configuration/ssh-zkfc-start.sh

```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
+
+# Copyright (c) Microsoft Corporation
+# All rights reserved.
+#
+# MIT License
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+# documentation files (the "Software"), to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
+# to permit persons to whom the Software is furnished to do so, subject to the following conditions:
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+# BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+# Change SSH config.
+
+chmod 600 /root/.ssh/config
+chmod 600 /root/.ssh/id_rsa
+sed -i "/^[^#]*UsePAM/ s/.*/#&/" /etc/ssh/sshd_config
+echo "UsePAM no" >> /etc/ssh/sshd_config
+echo "Port 2122" >> /etc/ssh/sshd_config
+
+# Restart SSH
+
+/etc/init.d/ssh restart
+
+
+# step 1 : Retry start zkfc, until successing.
+# step 2 : monitor zkfc. If crushed, restart it.
+# It is not important that you start the ZKFC and NameNode daemons in a particular order. On any given node you can start the ZKFC before or after its corresponding NameNode.
+while true; do
+
+    until hdfs zkfc
+    do
+        echo "retry to start zkfc"
+        kill `jps | grep "DFSZKFailoverController" | cut -d " " -f 1`
+    done
+
+    echo "zkfc starting finished"
+
+    while jps | grep -q "DFSZKFailoverController" ; do
+
+        sleep 300
+
+    done
+
+    echo "zkfc crushed, restart it"
```

Comment:
>ech [](start = 4, length = 3)

shall we sleep one sec here? what if "sleep 300" never execute? it will run in a busy loop?

Reply:
In which situation， will  "sleep 300" never execute?

Reply:
I think 1s is too frequent?